### ISSUE TITLE
PATCH /profile

### ISSUE THREAD

**Issue by h01d** [apavlidi/IT_API#48]
PATCH /profile returns code 502

Body:

 <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
    <html><head>
    <title>502 Proxy Error</title>
    </head><body>
    <h1>Proxy Error</h1>
    <p>The proxy server received an invalid
    response from an upstream server.<br />
    The proxy server could not handle the request <em><a href="/profile">PATCH&nbsp;/profile</a></em>.<p>
    Reason: <strong>Error reading from remote server</strong></p></p>
    <hr>
    <address>Apache/2.4.25 (Debian) Server at api.it.teithe.gr Port 443</address>
    </body></html>

**Comment by iamaldi** [comment]
@apavlidi Seems like a back-end resource is unreachable. Could you provide some server logs for more efficient bug tracking?

Thanks

**Comment by iamaldi** [comment]
@apavlidi @kvisnia Okay, I tracked down the issue to a `try-catch` block.

If you notice, inside the following `try-catch` block you're initializing `socialMediaExtra` by basically parsing whatever JSON values are coming through the `socialMediaExtras`https://github.com/apavlidi/IT_API/blob/aa3492d589ef1c2abf7828708a53f6e210872c23/routes/user/profile/function.js#L89-L94

The purpose of the `try-catch` block here is to see if any errors occurred during parsing. However, in case that the JSON parsing has been done successfully, you are forgetting to check if the actual value of `socialMediaExtra` has been initialized to some value or its still `null`. Thus, you're missing a necessary `if` statement that will need to perform that check.

Following execution, the application performs the JSON parsing, continues execution out of the `try-catch` block and since the `socialMediaExtra` has a value of `null` on line 100 you're trying to iterate over a `null` object.
https://github.com/apavlidi/IT_API/blob/aa3492d589ef1c2abf7828708a53f6e210872c23/routes/user/profile/function.js#L100-L115 

### Suggestion
My suggestion for a fix is to move all the code inside the try catch block. Then introduce a `if` statement to check if the `socialMediaExtra` has been initialized to a value other than `null` and if that statement is `true` then continue and perform the update.

In other words, the changes I propose can be seen below

```javascript
function updateExtraSocialMediaIfNecessary(ldapId, socialMediaExtras) {
  return new Promise(
    function (resolve, reject) {
      let socialMediaExtra
      try {
        socialMediaExtra = JSON.parse(socialMediaExtras)
        // Check if socialMediaExtra has been initialized to a value (other than null)
        if (socialMediaExtra) {
          // Find user from the database
          database.Profile.findOne({ ldapId: ldapId }, function (err, profile) {
            if (err || !profile) {
              reject(new PromiseError(2015, err))
            }

            // Iterate over the social media extras
            socialMediaExtra.forEach(function (value) {
              if (value.name && value.url) {
                if (extraSocialMediaExistsInProfile(profile, value)) {
                  findSocialMediaPosAndUpdate(profile, value)
                } else {
                  profile.socialMedia.socialMediaExtra.push(value)
                }
              }
            })

            // Save the user profile object, in other words update the profile in the database
            profile.save(function (err) {
              if (err) {
                reject(new PromiseError(2016, err))
              } else {
                resolve()
              }
            })
          })
        }
      } catch (err) {
        reject(new PromiseError(2014, err))
      }
    })
}
```

**Note**
Please note that an alternative `else` statement needs to be inserted. I left that out since I am not aware of which status / error code to use with the `reject(new PromiseError(2014, err))`

**Comment by iamaldi** [comment]
@apavlidi I could make the appropriate changes to the code and then create a PR but I will be unable to test the functionality myself. Could you do that after I create the PR and then merge if necessary?

**Comment by apavlidi** [comment]
@iamaldi Yes I could test it. You can also join our [slack channel](https://join.slack.com/t/itteithe/shared_invite/enQtNDE0NDk5NzExNjE2LTc1M2E2YTE0NTVmNzVjNjUyMjcxOTEyMTIxMTIwNzg0YzFkMmE4MWY4NThjMWRjNDk1ZjcxNTcxMWEyYWQ4Mzc) so we can discuss the project installation into your machine so you can test your own changes.
As for the change you purposed we recommend using the validation method we implemented as a middleware so we don`t have to validate input inside the the function that constructs the response. This implementation brings clarity to the code and separates concerns. Here is the use of the validationn method called **validateInput** .This function takes two arguments.First argument is the object that needs to be validated and the second argument is the schema that validates the object. We validate objects with the [Joi](https://www.npmjs.com/package/joi) module.
https://github.com/apavlidi/IT_API/blob/880b8b7e8d8891d46eab9f5fcf43b4d51358d96d/routes/user/profile/index.js#L18
